{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ea31c0b-1536-4df6-a25c-a493a781176f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Description\n",
    "print(\"\"\"This code uses Biopython to search DNA sequences from FASTA files(.fa) across the BLAST database.\"\"\")\n",
    "print(\"\\n\")\n",
    "print(\"Note: This program can return a maximum of 50 results per record.\")\n",
    "\n",
    "#Imports\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "import time\n",
    "\n",
    "#Takes path of FASTA file\n",
    "file_path=input(\"Please input the path of your FASTA file(.fa) here: \")\n",
    "\n",
    "#Takes the contents of the file into a string\n",
    "fasta_string=open(file_path).read()\n",
    "\n",
    "#Splits the file into records\n",
    "records=fasta_string.split(\">\")\n",
    "\n",
    "#Removes the blank string\n",
    "records.remove(records[0])\n",
    "\n",
    "#Re-adds the \">\" sign in each record\n",
    "for record in records:\n",
    "    records[records.index(record)] = \">\"+record\n",
    "\n",
    "#Asks which record should be searched\n",
    "queries = input(f\"\"\"Your file has {len(records)} records. If you want to search multiple records, put commas between each record number. Please input the number of the records you want to search: \"\"\").split(\",\")\n",
    "\n",
    "#Turns the strings to integers\n",
    "for query in queries:\n",
    "    queries[queries.index(query)]=int(query)\n",
    "\n",
    "#Requests what algorithm to use\n",
    "alg=input(\"Which algorithm do you want to use, blastn or megablast? \").lower()\n",
    "if alg==\"blastn\":\n",
    "    mb=None\n",
    "elif alg=="megablast:\n",
    "    mb=True\n",
    "\n",
    "#Searches using qblast\n",
    "for query in queries:\n",
    "    start_time=time.perf_counter()\n",
    "    print(\"\\n\")\n",
    "    print(\"\\n\")\n",
    "    print(\"\\n\")\n",
    "    print(f\"**** Searching Record Number {query} ****\")\n",
    "    result_handle=NCBIWWW.qblast(\"blastn\",\"nt\",records[query-1], megablast=mb)\n",
    "    \n",
    "    #Reads these results into the record\n",
    "    blast_record=NCBIXML.read(result_handle)\n",
    "\n",
    "    #Sets the e value threshold (Default BLAST parameters)\n",
    "    e_value_thresh=0.05\n",
    "    \n",
    "    #Checks if there are alignments to print\n",
    "    #Shows alignments\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < e_value_thresh:\n",
    "                print(\"****Alignment****\")\n",
    "                print(f\"Sequence: {alignment.title}\")\n",
    "                print(f\"Length: {alignment.length}\")\n",
    "                print(f\"E Value: {hsp.expect}\")\n",
    "                if hsp.query==hsp.sbjct:\n",
    "                    print(\"Exact Match!\")\n",
    "                    end_time=time.perf_counter()\n",
    "                    execution_time=abs(end_time-start_time)\n",
    "                    print(f\"Execution time: {round(execution_time)} seconds\")\n",
    "                    print(\"\\n\")\n",
    "                else:\n",
    "                    print(hsp.query)\n",
    "                    print(hsp.match)\n",
    "                    print(hsp.sbjct)\n",
    "                    end_time=time.perf_counter()\n",
    "                    execution_time=abs(end_time-start_time)\n",
    "                    print(f\"Execution time: {round(execution_time)} seconds\")\n",
    "                    print(\"\\n\")\n",
    "                    \n",
    "#Finishes results\n",
    "print(\"No other significant similarities found\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
